platform/ext_bitbucket.go

```diff
@@ -0,0 +1,216 @@
+	log "github.com/sirupsen/logrus"
+	"io"
+	"net/http"
+	"net/url"
+	"os"
+	"time"
+)
+
+const (
+	httpTimeout              = time.Second * 10
+	bitBucketAnnotationLimit = 1000
+	bitBucketReporter        = "JetBrains Qodana"
+	bitBucketAvatar          = "https://avatars.githubusercontent.com/u/139879315"
+	bitBucketReportFailed    = "FAILED"
+	bitBucketReportPassed    = "PASSED"
+	bitBucketReportType      = "BUG"
+	bitBucketAnnotationType  = "CODE_SMELL"
+
+	// https://developer.atlassian.com/cloud/bitbucket/rest/api-group-reports/#api-repositories-workspace-repo-slug-commit-commit-reports-reportid-annotations-annotationid-put-request
+	bitBucketHigh   = "HIGH"
+	bitBucketMedium = "MEDIUM"
+	bitBucketLow    = "LOW"
+	bitBucketInfo   = "INFO"
+
+	pipelineProxyURL = "http://localhost:29418"
+	pipeProxyURL     = "http://host.docker.internal:29418"
+)
+
+var (
+	toBitBucketSeverity = map[string]string{
+		sarifError:     bitBucketHigh,
+		sarifWarning:   bitBucketMedium,
+		sarifNote:      bitBucketLow,
+		qodanaCritical: bitBucketHigh,
+		qodanaHigh:     bitBucketMedium,
+		qodanaModerate: bitBucketMedium,
+		qodanaLow:      bitBucketLow,
+		qodanaInfo:     bitBucketInfo,
+	}
+)
+
+// sendBitBucketReport sends annotations to BitBucket Code Insights
+func sendBitBucketReport(annotations []bbapi.ReportAnnotation, toolName, cloudUrl, reportId string) error {
+	client, ctx := getBitBucketClient(), getBitBucketContext()
+	repoOwner, repoName, sha := getBitBucketRepoOwner(), getBitBucketRepoName(), getBitBucketCommit()
+	_, resp, err := client.
+		ReportsApi.CreateOrUpdateReport(ctx, repoOwner, repoName, sha, reportId).
+		Body(buildReport(toolName, annotations, cloudUrl)).
+		Execute()
+	if err = checkBitBucketApiError(err, resp, http.StatusOK); err != nil {
+		return fmt.Errorf("failed to create code insights report: %w", err)
+	}
+	totalAnnotations := len(annotations)
+	if totalAnnotations != 0 {
+		if totalAnnotations > bitBucketAnnotationLimit {
+			totalAnnotations = bitBucketAnnotationLimit
+			log.Debugf("Warning: Only first 1000 of %d annotations will be sent", len(annotations))
+		}
+		for i := 0; i < totalAnnotations; i += 100 {
+			j := i + 100
+			if j > totalAnnotations {
+				j = totalAnnotations
+			}
+			_, resp, err := client.ReportsApi.
+				BulkCreateOrUpdateAnnotations(ctx, repoOwner, repoName, sha, reportId).
+				Body(annotations[i:j]).
+				Execute()
+			if err = checkBitBucketApiError(err, resp, http.StatusOK); err != nil {
+				return fmt.Errorf("failed to create code insights annotations: %w", err)
+			}
+		}
+	}
+	return nil
+}
+
+// getBitBucketContext returns a context with BitBucket credentials (not required for runs in BitBucket Pipelines)
+func getBitBucketContext() context.Context {
+	ctx := context.Background()
+	user, password, token :=
+		os.Getenv("QD_BITBUCKET_USER"),
+		os.Getenv("QD_BITBUCKET_PASSWORD"),
+		os.Getenv("QD_BITBUCKET_TOKEN")
+	if user != "" && password != "" {
```

Comment:
Maybe it's more in the domain of the `bbapi` client itself, but I would expect us to avoid setting HTTP basic auth when a token is present

Reply:
It's used the same way as it is with reviewdog, even in the order of the operations. I think we could leave it for special cases here; unfortunately, it's not handled like that in the library